In [15]:
import nltk
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to C:\Users\Naman
[nltk_data]    |     Gupta\AppData\Roaming\nltk_data...
[nltk_data]    |   Package abc is already up-to-date!
[nltk_data]    | Downloading package alpino to C:\Users\Naman
[nltk_data]    |     Gupta\AppData\Roaming\nltk_data...
[nltk_data]    |   Package alpino is already up-to-date!
[nltk_data]    | Downloading package biocreative_ppi to C:\Users\Naman
[nltk_data]    |     Gupta\AppData\Roaming\nltk_data...
[nltk_data]    |   Package biocreative_ppi is already up-to-date!
[nltk_data]    | Downloading package brown to C:\Users\Naman
[nltk_data]    |     Gupta\AppData\Roaming\nltk_data...
[nltk_data]    |   Package brown is already up-to-date!
[nltk_data]    | Downloading package brown_tei to C:\Users\Naman
[nltk_data]    |     Gupta\AppData\Roaming\nltk_data...
[nltk_data]    |   Package brown_tei is already up-to-date!
[nltk_data]    | Downloading package cess_cat to C

True

**Input Doc**

In [16]:
doc1='''Web mining is the application of data mining techniques to discover patterns from the World Wide Web.
As the name proposes, this is information gathered by mining the web. It makes utilization of automated apparatuses 
to reveal and extricate data from servers and web2 reports, and it permits organizations to get to both organized and 
unstructured information from browser activities, server logs, website and link structure, page content and different sources.
The goal of Web structure mining is to generate structural summary about the Web site and Web page. Technically, Web content 
mining mainly focuses on the structure of inner-document, while Web structure mining tries to discover the link structure of 
the hyperlinks at the inter-document level. Based on the topology of the hyperlinks, Web structure mining will categorize 
the Web pages and generate the information, such as the similarity and relationship between different 
Web sites.Web structure mining can also have another direction – discovering the structure of Web document itself. 
This type of structure mining can be used to reveal the structure (schema) of Web pages, this would be good for navigation purpose 
and make it possible to compare/integrate Web page schemes.'''
doc2='''Web scraping, web harvesting, or web data extraction is data scraping used for extracting data from websites. 
Web scraping software may access the World Wide Web directly using the Hypertext Transfer Protocol, or through a web browser. 
While web scraping can be done manually by a software user, the term typically refers to automated processes implemented using 
a bot or web crawler. It is a form of copying, in which specific data is gathered and copied from the web, typically into a 
central local database or spreadsheet, for later retrieval or analysis.Web scraping a web page involves fetching it and 
extracting from it. Fetching is the downloading of a page (which a browser does when a user views a page). Therefore, web
crawling is a main component of web scraping, to fetch pages for later processing. Once fetched, then extraction can take place. 
The content of a page may be parsed, searched, reformatted, its data copied into a spreadsheet, and so on. Web scrapers 
typically take something out of a page, to make use of it for another purpose somewhere else. An example would be to find 
and copy names and phone numbers, or companies and their URLs, to a list (contact scraping).Web scraping is used for 
contact scraping, and as a component of applications used for web indexing, web mining and data mining, online price
change monitoring and price comparison, product review scraping (to watch the competition), gathering real estate 
listings, weather data monitoring, website change detection, research, tracking online presence and reputation, web mashup and
, web data integration.Web pages are built using text-based mark-up languages (HTML and XHTML), and frequently contain
a wealth of useful data in text form. However, most web pages are designed for human end-users and not for ease of automated use.'''

**With stemming**

In [17]:
from nltk.stem import LancasterStemmer
LS=LancasterStemmer()

In [18]:
Stem1formed=[LS.stem(i) for i in doc1.split()]
Stem1formed=" ".join(Stem1formed)
Stem1formed

'web min is the apply of dat min techn to discov pattern from the world wid web. as the nam proposes, thi is inform gath by min the web. it mak util of autom apparatus to rev and ext dat from serv and web2 reports, and it permit org to get to both org and unstruct inform from brows activities, serv logs, websit and link structure, pag cont and diff sources. the goal of web structure min is to gen structural sum about the web sit and web page. technically, web cont min main focus on the structure of inner-document, whil web structure min tri to discov the link structure of the hyperlink at the inter-document level. bas on the topolog of the hyperlinks, web structure min wil categ the web pag and gen the information, such as the simil and rel between diff web sites.web structure min can also hav anoth direct – discov the structure of web docu itself. thi typ of structure min can be us to rev the structure (schema) of web pages, thi would be good for navig purpos and mak it poss to compar

In [19]:
Stem2formed=[LS.stem(i) for i in doc2.split()]
Stem2formed=" ".join(Stem2formed)
Stem2formed

'web scraping, web harvesting, or web dat extract is dat scraping us for extract dat from websites. web scraping softw may access the world wid web direct us the hypertext transf protocol, or through a web browser. whil web scraping can be don man by a softw user, the term typ ref to autom process impl us a bot or web crawler. it is a form of copying, in which spec dat is gath and cop from the web, typ into a cent loc databas or spreadsheet, for lat retriev or analysis.web scraping a web pag involv fetch it and extract from it. fetch is the download of a pag (which a brows doe when a us view a page). therefore, web crawl is a main compon of web scraping, to fetch pag for lat processing. ont fetched, then extract can tak place. the cont of a pag may be parsed, searched, reformatted, it dat cop into a spreadsheet, and so on. web scrapers typ tak someth out of a page, to mak us of it for anoth purpos somewh else. an exampl would be to find and cop nam and phon numbers, or company and thei

In [20]:
from sklearn.feature_extraction.text import CountVectorizer
vect=CountVectorizer(binary=True)
corpusformed = [Stem1formed,Stem2formed]
vect.fit(corpusformed)

CountVectorizer(analyzer='word', binary=True, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)

In [21]:
vocab=vect.vocabulary_
len(vocab)

212

In [22]:
from sklearn.metrics.pairwise import cosine_similarity
similar = cosine_similarity(vect.transform([Stem1formed]).toarray(), vect.transform([Stem2formed]).toarray())
similar

array([[0.296923]])

**Without Stemming**


In [23]:
from sklearn.feature_extraction.text import CountVectorizer
vect=CountVectorizer(binary=True)

In [26]:
corpusnostem = [doc1,doc2]

In [27]:
vect.fit(corpusnostem)

CountVectorizer(analyzer='word', binary=True, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)

In [28]:
vocab=vect.vocabulary_
len(vocab)

220

In [29]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vect.transform([doc1]).toarray(), vect.transform([doc2]).toarray())
similarity

array([[0.26048132]])

**Results:**
**With Stemming:** 
  * Vocab = 212
  * Similarity = 0.296923

**Without Stemming:** 
  * Vocab = 220
  * Similarity = 0.26048132


**We can observe an increase in Cosine Similarity with Stemming. Also the vocabulary size decreases on stemming as multiple words with the same root are eliminated.**